firewood/src/diff.rs

```diff
@@ -0,0 +1,1744 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+#![allow(dead_code)]
+
+use std::cmp::Ordering;
+use std::fmt;
+use storage::{Child, Node, Path, SharedNode, TrieReader};
+
+use crate::db::BatchOp;
+use crate::merkle::{Key, Merkle, Value};
+use crate::stream::{
+    IterationNode, NodeStreamState, as_enumerated_children_iter, key_from_nibble_iter,
+};
+
+// State structs for different types of diff iteration nodes
+
+/// Two nodes that need to be compared against each other.
+/// This state occurs when we have matching children in both trees that need to be processed,
+/// or when we're starting the comparison at the root level with nodes from both trees.
+#[derive(Debug)]
+struct UnvisitedPairState {
+    node_left: SharedNode,
+    node_right: SharedNode,
+}
+
+/// A node that exists only in the left tree (tree1) and needs to be processed for deletion.
+/// This state occurs when comparing children and the left tree has a child at a position
+/// where the right tree doesn't, or when the entire left subtree needs to be deleted.
+#[derive(Debug)]
+struct UnvisitedLeftState {
+    node: SharedNode,
+}
+
+/// A node that exists only in the right tree (tree2) and needs to be processed for addition.
+/// This state occurs when comparing children and the right tree has a child at a position
+/// where the left tree doesn't, or when the entire right subtree needs to be added.
+#[derive(Debug)]
+struct UnvisitedRightState {
+    node: SharedNode,
+}
+
+/// Two branch nodes whose children are being compared.
+/// This state occurs when we have branch nodes from both trees and we need to iterate
+/// through their children to find differences between the subtrees.
+struct VisitedPairState {
+    children_iter_left: Box<dyn Iterator<Item = (u8, Child)> + Send>,
+    children_iter_right: Box<dyn Iterator<Item = (u8, Child)> + Send>,
+}
+
+impl fmt::Debug for VisitedPairState {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("VisitedPairState")
+            .field("children_iter_left", &"<iterator>")
+            .field("children_iter_right", &"<iterator>")
+            .finish()
+    }
+}
```

Comment:
We can use [`derivative`](https://mcarton.github.io/rust-derivative/latest/index.html) to clean up these manually implemented `Debug`s.